driver-core/src/main/com/mongodb/operation/DistinctOperation.java

```diff
@@ -18,30 +18,50 @@
 
 import com.mongodb.Function;
 import com.mongodb.MongoNamespace;
+import com.mongodb.async.AsyncBatchCursor;
 import com.mongodb.async.SingleResultCallback;
+import com.mongodb.binding.AsyncConnectionSource;
 import com.mongodb.binding.AsyncReadBinding;
+import com.mongodb.binding.ConnectionSource;
 import com.mongodb.binding.ReadBinding;
-import org.bson.BsonArray;
+import com.mongodb.connection.Connection;
+import com.mongodb.connection.QueryResult;
 import org.bson.BsonDocument;
+import org.bson.BsonDocumentReader;
+import org.bson.BsonDocumentWrapper;
 import org.bson.BsonString;
+import org.bson.BsonValue;
+import org.bson.Document;
+import org.bson.codecs.Decoder;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.DocumentCodec;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.concurrent.TimeUnit;
 
 import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.async.ErrorHandlingResultCallback.errorHandlingCallback;
 import static com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol;
 import static com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocolAsync;
 import static com.mongodb.operation.DocumentHelper.putIfNotNull;
 import static com.mongodb.operation.DocumentHelper.putIfNotZero;
+import static com.mongodb.operation.OperationHelper.releasingCallback;
+import static com.mongodb.operation.OperationHelper.withConnection;
 
 /**
- * Finds the distinct values for a specified field across a single collection. This returns an array of the distinct values.
- * 
+ * Finds the distinct values for a specified field across a single collection.
+ *
  * <p>When possible, the distinct command uses an index to find documents and return values.</p>
  *
  * @mongodb.driver.manual reference/command/distinct Distinct Command
  * @since 3.0
  */
-public class DistinctOperation implements AsyncReadOperation<BsonArray>, ReadOperation<BsonArray> {
+public class DistinctOperation implements AsyncReadOperation<AsyncBatchCursor<Object>>, ReadOperation<BatchCursor<Object>> {
+    private static final String VALUES = "values";
+    private static final Decoder<Document> DECODER = new DocumentCodec();
```

Comment:
OK, now I see what you did.  Any reason not to pass the Decoder into the constructor?  This would be the first time we reference the Document class in any operation.